server/src/main/java/org/apache/iotdb/db/query/executor/LastQueryExecutor.java

```diff
@@ -0,0 +1,216 @@
+    Filter timeFilter = null;
+    IAggregateReader seriesReader = new SeriesAggregateReader(
+            seriesPath, tsDataType, context, QueryResourceManager.getInstance()
+            .getQueryDataSource(seriesPath, context, timeFilter), timeFilter, null);
+
+    long maxTime = Long.MIN_VALUE;
+    while (seriesReader.hasNextChunk()) {
+      // cal by chunk statistics
+      if (seriesReader.canUseCurrentChunkStatistics()) {
+        Statistics chunkStatistics = seriesReader.currentChunkStatistics();
+        if (chunkStatistics.getEndTime() > maxTime) {
+          maxTime = chunkStatistics.getEndTime();
+          queryResult.setPairResult(maxTime, chunkStatistics.getLastValue(), tsDataType);
+        }
+        seriesReader.skipCurrentChunk();
+        continue;
+      }
+      while (seriesReader.hasNextPage()) {
+        //cal by page statistics
+        if (seriesReader.canUseCurrentPageStatistics()) {
+          Statistics pageStatistic = seriesReader.currentPageStatistics();
+          if (pageStatistic.getEndTime() > maxTime) {
+            maxTime = pageStatistic.getEndTime();
+            queryResult.setPairResult(maxTime, pageStatistic.getLastValue(), tsDataType);
+          }
+          seriesReader.skipCurrentPage();
+          continue;
+        }
+        // cal by page data
+        while (seriesReader.hasNextOverlappedPage()) {
+          BatchData nextOverlappedPageData = seriesReader.nextOverlappedPage();
+          int maxIndex = nextOverlappedPageData.length() - 1;
+          if (maxIndex < 0) {
+            continue;
+          }
+          long time = nextOverlappedPageData.getTimeByIndex(maxIndex);
+          if (time > maxTime) {
+            maxTime = time;
+            queryResult.setPairResult(maxTime, nextOverlappedPageData.getValueInTimestamp(time), tsDataType);
+          }
+          nextOverlappedPageData.resetBatchData();
+        }
+      }
+    }
+    if (queryResult.hasResult())
+      node.setCachedLast(queryResult.getPairResult());
+    return queryResult;
+  }
+
+  /**
+   * using last result data list construct QueryDataSet.
+   *
+   * @param lastQueryResultList last result list
+   */
+  private RowRecord constructLastRowRecord(List<LastQueryResult> lastQueryResultList) {
+    long maxTime = Long.MIN_VALUE;
+    for (LastQueryResult lastPair : lastQueryResultList) {
+      if (lastPair.hasResult() && lastPair.getTimestamp() > maxTime)
+        maxTime = lastPair.getTimestamp();
+    }
+
+    RowRecord resultRecord = new RowRecord(maxTime);
+    for (int i = 0; i < lastQueryResultList.size(); i++) {
+      TSDataType dataType = dataTypes.get(i);
+      LastQueryResult lastPair = lastQueryResultList.get(i);
+      if (lastPair.hasResult() && lastPair.getTimestamp() == maxTime)
+        resultRecord.addField(lastPair.getValue(), dataType);
+      else
+        resultRecord.addField(null, dataType);
+    }
+
+    return resultRecord;
+  }
+
+  class LastQueryResult {
```

Comment:
I think this class is not needed, you could use TimeValuePair directly unless you store the path of a series in this class.

Reply:
Removed in the new version